Expose public equality comparison and sameness public API.

crysehillmes opened the pull request:
- add `JS_IsLooselyEqual`
- add `JS_IsStrictlyEqual`
- add `JS_IsSameValue`
- add `JS_IsSameValueZero`

crysehillmes committed:
Expose public equality comparison and sameness public API.
- add `JS_IsLooselyEqual`
- add `JS_IsStrictlyEqual`
- add `JS_IsSameValue`
- add `JS_IsSameValueZero`

saghul commented on quickjs.c:
```diff
@@ -51816,6 +51816,35 @@ void JS_AddPerformance(JSContext *ctx)
+/* Equality comparisons and sameness */
+JS_BOOL JS_IsLooselyEqual(JSContext *ctx, JSValue op1, JSValue op2)
+{
+    JSValue* sp = js_malloc(ctx, sizeof(JSValue) * 3);
```
I think you should be able to do this by allocating the array on the stack.

crysehillmes replied:
Thanks for your review suggestions.
I have completed the modifications in the new commit.

chqrlie commented on quickjs.c:
```diff
@@ -51816,6 +51816,35 @@ void JS_AddPerformance(JSContext *ctx)
+    js_free(ctx, sp);
+    return res;
+}
+
```
The return type should be `int` as `-1` is returned if an exception is thrown by a type conversion.
Do not allocate memory for this, use a local array `JSValue sp[3]`
No need for `JSValue isEqual` intermediary variable, `JS_ToBool` should be `JS_ToBoolFree`, but since we know the value is a `JS_TAG_BOOL`, just use `JS_VALUE_GET_BOOL(sp[0])`

Here is my suggestion:
```
int JS_IsLooselyEqual(JSContext *ctx, JSValue op1, JSValue op2)
{
    JSValue sp[3];
    sp[0] = js_dup(op1);
    sp[1] = js_dup(op2);
    if (js_eq_slow(ctx, &sp[2], 0))
        return -1;
    return JS_VALUE_GET_BOOL(sp[0]);
}
```

chqrlie commented on quickjs.h:
```diff
@@ -335,6 +335,12 @@ JS_EXTERN void JS_AddIntrinsicBigInt(JSContext *ctx);
 JS_EXTERN void JS_AddPerformance(JSContext *ctx);
 
+/* for equality comparisons and sameness */
+JS_EXTERN JS_BOOL JS_IsLooselyEqual(JSContext *ctx, JSValue op1, JSValue op2);
```
return type `int` for `JS_IsLooselyEqual`

bnoordhuis commented on quickjs.c:
```diff
@@ -51816,6 +51816,35 @@ void JS_AddPerformance(JSContext *ctx)
+    JSValue* sp = js_malloc(ctx, sizeof(JSValue) * 3);
+    sp[0] = js_dup(op1);
+    sp[1] = js_dup(op2);
+    int res = js_eq_slow(ctx, &(sp[2]), 0);
```
```suggestion
    int res = js_eq_slow(ctx, endof(sp), 0);
```
And change sp to `JSValue sp[2];` - right now, it's heap-allocated and allocates one element too many.

chqrlie replied:
The internal API `js_eq_slow` is really not meant to be called outside the interpreter loop. We should probably add
```
static int js_eq(JSContext *ctx, JSValue op1, JSValue op2);
```
that consumes its operands and returns the boolean value and -1 for exceptions. We would use it from `js_eq_slow` just like the handlers for other relational operators.

bnoordhuis commented on quickjs.c:
```diff
@@ -51816,6 +51816,35 @@ void JS_AddPerformance(JSContext *ctx)
+    int res = js_eq_slow(ctx, &(sp[2]), 0);
+    if (res == -1) return res;
+    JSValue isEqual = sp[0];
+    res = JS_ToBool(ctx, isEqual);
```
```suggestion
    res = JS_ToBoolFree(ctx, sp[0]);
```

bnoordhuis commented on quickjs.h:
```diff
@@ -335,6 +335,12 @@ JS_EXTERN void JS_AddIntrinsicBigInt(JSContext *ctx);
+JS_EXTERN JS_BOOL JS_IsLooselyEqual(JSContext *ctx, JSValue op1, JSValue op2);
+JS_EXTERN JS_BOOL JS_IsStrictlyEqual(JSContext *ctx, JSValue op1, JSValue op2);
+JS_EXTERN JS_BOOL JS_IsSameValue(JSContext *ctx, JSValue op1, JSValue op2);
+JS_EXTERN JS_BOOL JS_IsSameValueZero(JSContext *ctx, JSValue op1, JSValue op2);
```
- I'd name the first two JS_IsEqual and JS_IsStrictEqual

- can you add a comment explaining the difference between IsSameValue and IsSameValueZero? The latter considers -0 and +0 equal but my guess is that isn't obvious from the name (I had to double-check myself)

crysehillmes replied:
Thanks for your review suggestions.
I use those names from [Equality comparisons and sameness](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality) of MDN.
But `JS_IsEqual` and `JS_IsStrictEqual` seems more convenient so I've changed in new commit.

chqrlie replied:
> Thanks for your review suggestions. I use those names from [Equality comparisons and sameness](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality) of MDN. But `JS_IsEqual` and `JS_IsStrictEqual` seems more convenient so I've changed in new commit.

Using the same terms as ECMA seems a good idea.  Using a comment to clarify definitely helps, along with a reference to the standard.

crysehillmes committed:
Addressing the feedback from the code review:
- Rename `JS_IsLooselyEqual` to `JS_IsEqual`
- Rename `JS_IsStrictlyEqual` to `JS_IsStrictEqual`
- Use stack instead of heap in `JS_IsEqual`

chqrlie commented on quickjs.c:
```diff
@@ -51816,6 +51816,31 @@ void JS_AddPerformance(JSContext *ctx)
 }
 
+/* Equality comparisons and sameness */
+JS_BOOL JS_IsEqual(JSContext *ctx, JSValue op1, JSValue op2)
```
The return type **must be `int`** because there are 3 possible return values: -1, 0 and 1

chqrlie commented on quickjs.h:
```diff
@@ -335,6 +335,13 @@ JS_EXTERN void JS_AddIntrinsicBigInt(JSContext *ctx);
 JS_EXTERN void JS_AddPerformance(JSContext *ctx);
 
+/* for equality comparisons and sameness */
+JS_EXTERN JS_BOOL JS_IsEqual(JSContext *ctx, JSValue op1, JSValue op2);
```
Please change the return type for `JS_IsEqual` to `int`

crysehillmes replied:
Thanks for your review suggestions, I've changed the return type for `JS_IsEqual` to `int`.

crysehillmes committed:
Addressing the feedback from the code review:
- Use `int` instead of `JS_BOOL` as `JS_IsEqual` return type
- Use `JS_VALUE_GET_BOOL` instead of `JS_ToBoolFree` in `JS_IsEqual`

bnoordhuis commented on quickjs.h:
```diff
@@ -335,6 +335,13 @@ JS_EXTERN void JS_AddIntrinsicBigInt(JSContext *ctx);
 
+/* for equality comparisons and sameness */
+JS_EXTERN int JS_IsEqual(JSContext *ctx, JSValue op1, JSValue op2);
+JS_EXTERN JS_BOOL JS_IsStrictEqual(JSContext *ctx, JSValue op1, JSValue op2);
+JS_EXTERN JS_BOOL JS_IsSameValue(JSContext *ctx, JSValue op1, JSValue op2);
+/* Similar to same-value equality, but +0 and -0 are considered equal. */
+JS_EXTERN JS_BOOL JS_IsSameValueZero(JSContext *ctx, JSValue op1, JSValue op2);
```
Shouldn't these be s/JS_BOOL/int/ as well?

bnoordhuis replied:
Ah wait, only js_eq_slow is fallible, the others aren't (somewhat to my surprise)
